providers/implementations/keymgmt/ml_kem_kmgmt.c

```diff
@@ -0,0 +1,518 @@
+    }
+    /* An unlikely failure mode is the parameter having some unexpected type */
+    if (!OSSL_PARAM_get_octet_string_ptr(p, &pubenc, &publen))
+        return 0;
+
+    if (publen != v->pubkey_bytes) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY);
+        return 0;
+    }
+
+    return ossl_ml_kem_parse_public_key(pubenc, publen, key);
+}
+
+static int ml_kem_gen_set_params(void *vgctx, const OSSL_PARAM params[])
+{
+    PROV_ML_KEM_GEN_CTX *gctx = vgctx;
+    const OSSL_PARAM *p;
+
+    if (gctx == NULL)
+        return 0;
+    if (ossl_param_is_empty(params))
+        return 1;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PROPERTIES);
+    if (p != NULL) {
+        if (p->data_type != OSSL_PARAM_UTF8_STRING)
+            return 0;
+        OPENSSL_free(gctx->propq);
+        if ((gctx->propq = OPENSSL_strdup(p->data)) == NULL)
+            return 0;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_ML_KEM_SEED);
+    if (p != NULL) {
+        size_t len = ML_KEM_SEED_BYTES;
+
+        gctx->seed = gctx->seedbuf;
+        if (OSSL_PARAM_get_octet_string(p, (void **)&gctx->seed, len, &len)
+            && len == ML_KEM_SEED_BYTES)
+            return 1;
+
+        /* Possibly, but less likely wrong data type */
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_SEED_LENGTH);
+        gctx->seed = NULL;
+        return 0;
+    }
+    return 1;
+}
+
+static void *ml_kem_gen_init(void *provctx, int selection,
+                             const OSSL_PARAM params[], int variant)
+{
+    PROV_ML_KEM_GEN_CTX *gctx = NULL;
+
+    /*
+     * We can only generate private keys, check that the selection is
+     * appropriate.
+     */
+    if (!ossl_prov_is_running()
+        || (selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0
+        || (gctx = OPENSSL_zalloc(sizeof(*gctx))) == NULL)
+        return NULL;
+
+    if (provctx != NULL)
+        gctx->libctx = PROV_LIBCTX_OF(provctx);
+    gctx->variant = variant;
+
+    if (ml_kem_gen_set_params(gctx, params))
+        return gctx;
+
+    ml_kem_gen_cleanup(gctx);
+    return NULL;
+}
+
+static const OSSL_PARAM *ml_kem_gen_settable_params(ossl_unused void *vgctx,
+                                                    ossl_unused void *provctx)
+{
+    static OSSL_PARAM settable[] = {
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_ML_KEM_SEED, NULL, 0),
+        OSSL_PARAM_END
+    };
+    return settable;
+}
+
+static void *ml_kem_gen(void *vgctx, OSSL_CALLBACK *osslcb, void *cbarg)
+{
+    PROV_ML_KEM_GEN_CTX *gctx = vgctx;
+    ML_KEM_KEY *key;
+    uint8_t *nopub = NULL;
+    uint8_t *seed = gctx->seed;
+    size_t slen = seed == NULL ? 0 : ML_KEM_SEED_BYTES;
+    int genok = 0;
+
+    if (gctx == NULL
+        || (key = ml_kem_new(gctx->libctx, gctx->propq, gctx->variant)) == NULL)
+        return NULL;
+
+    genok = seed != NULL ?
+        ossl_ml_kem_genkey_seed(seed, slen, nopub, 0, key) :
+        ossl_ml_kem_genkey_rand(seed, slen, nopub, 0, key);
```

Comment:
Nit - a proper formatting according to our style is:
```suggestion
    genok = seed != NULL
            ? ossl_ml_kem_genkey_seed(seed, slen, nopub, 0, key)
            : ossl_ml_kem_genkey_rand(seed, slen, nopub, 0, key);
```

Also I would not pass seed in the second call when we know that it is NULL. I would explicitly pass NULL as the argument.